src/os/bluestore/bluefs_types.h

```diff
@@ -42,16 +48,66 @@ struct bluefs_fnode_delta_t {
   uint64_t offset; // Contains offset in file of extents.
                    // Equal to 'allocated' when created.
                    // Used for consistency checking.
+
+  // only relevant in case of wal node
+  uint64_t wal_limit;
+  uint64_t wal_size;
+  uint8_t type; // bluefs_node_type. don't encode this.
+
+
   mempool::bluefs::vector<bluefs_extent_t> extents;
 
-  DENC(bluefs_fnode_delta_t, v, p) {
-    DENC_START(1, 1, p);
+  DENC_HELPERS
+
+  void bound_encode(size_t& p) const {
+    _denc_friend(*this, p);
+  }
+  void encode(ceph::buffer::list::contiguous_appender& p) const {
+    DENC_DUMP_PRE(bluefs_fnode_t);
+    _denc_friend(*this, p);
+  }
+  void decode(ceph::buffer::ptr::const_iterator& p) {
+    DENC_START(2, 2, p);
+    denc_varint(ino, p);
+    denc_varint(size, p);
+    denc(mtime, p);
+    denc(offset, p);
+    denc(extents, p);
+
+    denc(type, p);
+    if (struct_v >= 2 && type == WAL_V2) {
+      ceph_assert(struct_v == 2 && struct_compat == 2);
+      denc(wal_limit, p);
+      denc(wal_size, p);
+    }
+    DENC_FINISH(p);
+
+  }
+
+  template<typename T, typename P>
+  friend std::enable_if_t<std::is_same_v<bluefs_fnode_delta_t, std::remove_const_t<T>>>
+  _denc_friend(T& v, P& p) {
+    uint8_t version = 1, compat = 1;
+    if (v.type == WAL_V2) {
+      version = 2;
+      compat = 2;
+    }
+    DENC_START(version, compat, p);
+
     denc_varint(v.ino, p);
     denc_varint(v.size, p);
     denc(v.mtime, p);
     denc(v.offset, p);
     denc(v.extents, p);
+
+    denc(v.type, p);
```

Comment:
IMO this is wrong (as you update structure without version increase) and it could be implemented differently (and simpler):
- On encode struct's version/compat to be set to 2. And all 3 new fields to be encoded. On decode - if version is 1 then type to be set to v1, overwise - to be decoded.

Reply:
done I think